File: learnLeetCode/Day10/Part3/shadow_network.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
using namespace std;

class UnionFind {
    map<int, int> parent;
public:
    int find(int x) {
        if (parent.find(x) == parent.end()) parent[x] = x;
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    bool unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;
        parent[py] = px;
        return true;
    }
};

int computeCostWithRelays(int n, const vector<vector<int>>& edges, const set<int>& relays) {
    map<pair<int,int>, int> edgeMap;
    for (auto &e : edges) {
        int u = e[0], v = e[1], w = e[2];
        auto key = make_pair(min(u,v), max(u,v));
        if (edgeMap.find(key) == edgeMap.end() || edgeMap[key] > w) {
            edgeMap[key] = w;
        }
    }

    int cost = 0;
    set<int> nonRelays;
    for (int i = 0; i < n; ++i) {
        if (relays.find(i) == relays.end()) nonRelays.insert(i);
    }

    // Connect each non-relay to cheapest relay
    for (int node : nonRelays) {
        int minCost = INT_MAX;
        for (int relay : relays) {
            auto key = make_pair(min(node, relay), max(node, relay));
            if (edgeMap.find(key) != edgeMap.end()) {
                minCost = min(minCost, edgeMap[key]);
            }
        }
        if (minCost == INT_MAX) return INT_MAX;
        cost += minCost;
    }

    // MST among relays
    vector<vector<int>> relayEdges;
    vector<int> relayList(relays.begin(), relays.end());
    for (size_t i = 0; i < relayList.size(); ++i) {
        for (size_t j = i+1; j < relayList.size(); ++j) {
            int u = relayList[i], v = relayList[j];
            auto key = make_pair(min(u,v), max(u,v));
            if (edgeMap.find(key) != edgeMap.end()) {
                relayEdges.push_back({u, v, edgeMap[key]});
            }
        }
    }

    sort(relayEdges.begin(), relayEdges.end(), [](auto &a, auto &b) { return a[2] < b[2]; });
    UnionFind uf;
    for (auto &e : relayEdges) {
        if (uf.unite(e[0], e[1])) {
            cost += e[2];
        }
    }

    return cost;
}

void generateCombinations(int n, int k, int start, vector<int>& current, vector<vector<int>>& result) {
    if ((int)current.size() == k) {
        result.push_back(current);
        return;
    }
    for (int i = start; i < n; ++i) {
        current.push_back(i);
        generateCombinations(n, k, i+1, current, result);
        current.pop_back();
    }
}

int shadowNetwork(int n, const vector<vector<int>>& edges, int k) {
    if (k >= n) {
        // MST
        vector<vector<int>> sortedEdges = edges;
        sort(sortedEdges.begin(), sortedEdges.end(), [](auto &a, auto &b) { return a[2] < b[2]; });
        UnionFind uf;
        int cost = 0;
        for (auto &e : sortedEdges) {
            if (uf.unite(e[0], e[1])) cost += e[2];
        }
        return cost;
    }

    vector<vector<int>> combos;
    vector<int> current;
    generateCombinations(n, k, 0, current, combos);

    int minCost = INT_MAX;
    for (auto &combo : combos) {
        set<int> relaySet(combo.begin(), combo.end());
        int cost = computeCostWithRelays(n, edges, relaySet);
        minCost = min(minCost, cost);
    }

    return (minCost == INT_MAX) ? -1 : minCost;
}

int main() {
    int res = shadowNetwork(5, {{0,1,4},{0,2,3},{1,3,2},{2,4,6},{3,4,1}}, 2);
    cout << "Day10 Part3 shadow_network.cpp result for example: " << res << endl;

    // Simple test
    if (shadowNetwork(3, {{0,1,1},{1,2,1},{0,2,2}}, 1) != 2) return 1;

    cout << "Day10 Part3 shadow_network.cpp: basic tests passed" << endl;
    return 0;
}

File: Day10/Part3/shadow_network.py
"""
Day 10 — Part 3: The Shadow Network (k relay nodes + ≤2 hop)

Given n nodes, edges, and k (max relay nodes), find minimum cost to connect
all nodes such that:
- Only k nodes can act as relay nodes (intermediate nodes in paths).
- Any two nodes can communicate within ≤2 hops, using only relay nodes as intermediates.

Approach (Heuristic for small graphs):
1. Select k nodes as relays (try different combinations or use greedy heuristic).
2. Connect all non-relay nodes to at least one relay (direct edge).
3. Ensure relays are interconnected such that any relay-to-relay path ≤1 hop (complete graph among relays, or MST).
4. Compute total cost and pick minimum across different relay selections.

This is NP-hard; for small n, we can try different k-node subsets.
For larger n, use heuristics like selecting high-degree nodes as relays.

Time: O(C(n,k) * E) for brute force, or O(n^2) for greedy heuristic.
Space: O(V + E)
"""
from typing import List
from itertools import combinations


def compute_cost_with_relays(n: int, edges: List[List[int]], relays: set) -> int:
    """
    Compute minimum cost to connect all nodes using given relay nodes.
    All non-relay nodes must connect directly to at least one relay.
    Relays must be fully connected among themselves.
    """
    edge_map = {}
    for u, v, w in edges:
        edge_map[(min(u,v), max(u,v))] = min(edge_map.get((min(u,v), max(u,v)), float('inf')), w)

    cost = 0
    non_relays = set(range(n)) - relays

    # Connect each non-relay to cheapest relay
    for node in non_relays:
        min_cost = float('inf')
        for relay in relays:
            key = (min(node, relay), max(node, relay))
            if key in edge_map:
                min_cost = min(min_cost, edge_map[key])
        if min_cost == float('inf'):
            return float('inf')  # Can't connect
        cost += min_cost

    # Connect relays among themselves (MST among relays)
    relay_list = list(relays)
    relay_edges = []
    for i in range(len(relay_list)):
        for j in range(i+1, len(relay_list)):
            u, v = relay_list[i], relay_list[j]
            key = (min(u,v), max(u,v))
            if key in edge_map:
                relay_edges.append((u, v, edge_map[key]))

    # Kruskal on relay edges
    from collections import defaultdict
    parent = {r: r for r in relays}
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    def union(x, y):
        px, py = find(x), find(y)
        if px == py:
            return False
        parent[py] = px
        return True

    relay_edges.sort(key=lambda e: e[2])
    for u, v, w in relay_edges:
        if union(u, v):
            cost += w

    return cost


def shadow_network(n: int, edges: List[List[int]], k: int) -> int:
    """
    Returns minimum cost to connect all nodes with k relay nodes and ≤2 hop constraint.
    Brute force: try all combinations of k nodes as relays.
    """
    if k >= n:
        # All nodes can be relays; simple MST
        from collections import defaultdict
        uf_parent = list(range(n))
        def find(x):
            if uf_parent[x] != x:
                uf_parent[x] = find(uf_parent[x])
            return uf_parent[x]
        def union(x, y):
            px, py = find(x), find(y)
            if px == py:
                return False
            uf_parent[py] = px
            return True
        edges_sorted = sorted(edges, key=lambda e: e[2])
        cost = 0
        for u, v, w in edges_sorted:
            if union(u, v):
                cost += w
        return cost

    min_cost = float('inf')
    for relay_combo in combinations(range(n), k):
        relay_set = set(relay_combo)
        cost = compute_cost_with_relays(n, edges, relay_set)
        min_cost = min(min_cost, cost)

    return min_cost if min_cost != float('inf') else -1


if __name__ == "__main__":
    # Example from problem
    n = 5
    edges = [[0,1,4],[0,2,3],[1,3,2],[2,4,6],[3,4,1]]
    k = 2
    result = shadow_network(n, edges, k)
    print(f"Day10 Part3 shadow_network.py result for example: {result}")
    # Expected: 10 (activate nodes [1,3], use edges [0,1,4],[1,3,2],[3,4,1],[0,2,3])

    # Simple test
    assert shadow_network(3, [[0,1,1],[1,2,1],[0,2,2]], 1) == 2  # relay at node 1

    print("Day10 Part3 shadow_network.py: basic tests passed")
